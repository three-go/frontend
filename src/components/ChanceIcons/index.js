import React from "react";

import { View, StyleSheet } from "react-native";
import Icon from "react-native-vector-icons/MaterialCommunityIcons";

import { iconNames, iconSizes, colors } from "../../common";

const MAX_CHANCE = 3;

const ChanceIcons = ({ chance }) => {
  return (
    <View style={styles.heartWrapper}>
      {createChanceIcons(MAX_CHANCE, chance).map((value, index) => (
        <Icon
          key={index}
          name={value}
          size={iconSizes.headerChance}
          color={colors.red}
        />
      ))}
    </View>
  );
};

const createChanceIcons = (maxChance, restChance) => {
  const result = [];
  let restCount = restChance;

  for (let i = 0; i < maxChance; i++) {
    if (restCount > 0) {
      --restCount;
      result.push(iconNames.heart);
    } else {
      result.push(iconNames.brokenHeart);
    }
  }

  return result;
};

const styles = StyleSheet.create({
  heartWrapper: {
    flexDirection: "row",
  },
});

export default ChanceIcons;
