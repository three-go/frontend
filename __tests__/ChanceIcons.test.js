import React from "react";
import { render } from "@testing-library/react-native";
import "@testing-library/react-native";
import ChanceIcons from "../src/components/ChanceIcons";

const getRandomChanceCount = () => {
  return Math.floor(Math.random() * 4);
};

const getChanceIconsId = (rest) => {
  const MAX_CHANCE = 3;
  const BROKEN_HEART = "heart-broken-outline";
  const HEART = "heart";
  const result = [];

  for (let i = 0; i < MAX_CHANCE; i++) {
    if (rest > 0) {
      result.push(HEART);
      rest--;
    } else {
      result.push(BROKEN_HEART);
    }
  }

  return result;
};

it("Component must have 3 chance icons.", () => {
  const chanceCount = 3;
  const { queryAllByTestId, toJSON } = render(
    <ChanceIcons chance={chanceCount} />
  );
  const chanceIcons = toJSON();

  expect(queryAllByTestId("heart").length).toBe(chanceCount);
  expect(chanceIcons).toMatchSnapshot();
  expect(chanceIcons).toBeTruthy();
});

it("The icon should change according to the number of chance remaining.", () => {
  const chanceCount = getRandomChanceCount();
  const expectedResult = getChanceIconsId(chanceCount);
  const { toJSON } = render(<ChanceIcons chance={chanceCount} />);
  const children = toJSON().children;

  for (let i = 0; i < 3; i++) {
    expect(children[i].props.testID).toBe(expectedResult[i]);
  }
});
