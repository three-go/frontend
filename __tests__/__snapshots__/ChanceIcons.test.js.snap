// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component must have 3 chance icons. 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "#f45a5a",
          "fontSize": 28,
        },
        undefined,
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="heart"
  >
    󰋑
  </Text>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "#f45a5a",
          "fontSize": 28,
        },
        undefined,
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="heart"
  >
    󰋑
  </Text>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "#f45a5a",
          "fontSize": 28,
        },
        undefined,
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="heart"
  >
    󰋑
  </Text>
</View>
`;
